lib/src/components/NxFieldset/stateHelpers.ts

```diff
@@ -0,0 +1,144 @@
+ */
+export function radioGroupUserInput(
+  value: string | null,
+  validator?: RadioValidator
+): RadioStateProps {
+  return {
+    value,
+    isPristine: false,
+    validationErrors: validator ? validator(value) : null
+  };
+}
+
+/**
+ * A react hook to aid in managing the state of a radio group. Encapsulates tracking of pristine state and
+ * validation errors
+ */
+export function useRadioGroupState(
+  initialValue?: string,
+  validator?: RadioValidator
+): [RadioStateProps, RadioSetter] {
+  const [state, setState] = useState(radioGroupInitialState(initialValue, validator)),
+      setter = (v: string | null) => { setState(radioGroupUserInput(v, validator)); };
+
+  return [state, setter];
+}
+
+/**
+ * Create a CheckboxStateProps representing the initial state of a group of checkboxes based on the
+ * specified initial values, and validated according to the specified validator
+ */
+export function checkboxGroupInitialState(
+  values: string[] = [],
+  validator?: CheckboxValidator
+): CheckboxStateProps {
+  return {
+    values,
+    isPristine: true,
+    validationErrors: validator ? validator(values) : null
+  };
+}
+
+/**
+ * Create a CheckboxStateProps representing a non-initial state of a group of checkboxes based on the
+ * specified value, and validated according to the specified validator
+ */
+export function checkboxGroupUserInput(
+  { values }: CheckboxStateProps,
+  toggledValue: string,
+  validator?: CheckboxValidator
+): CheckboxStateProps {
+  const newValues = values.includes(toggledValue) ?
+    without([toggledValue], values) :
+    append(toggledValue, values);
+
+  return {
+    values: newValues,
+    isPristine: false,
+    validationErrors: validator ? validator(newValues) : null
+  };
+}
+
+// Helper to make TypeScript happy with reducing useToggle's 3-tuple to a 2-tuple
+function _useToggle(initialValue: boolean): CheckboxState {
+  const [state, toggle] = useToggle(initialValue);
+  return [state, toggle];
+}
+
+/**
+ * A react hook to aid in managing the state of a checkbox group. Takes an object mapping state names (i.e. checkbox
+ * names) to the initial values for those checkboxes, along with an optional validator function, and returns
+ * an object containing React state values and setters (as if from useState) for each checkbox along with isPristine and
+ * validationErrors
+ */
+export function useCheckboxGroupState<K extends string>(
+  initialValues: CheckboxInitValues<K>,
+  validator?: CheckboxValidator
+): CheckboxGroupHookReturnValue<K> {
+  type Pair = [K, CheckboxState];
+
+  function wrapState([state, toggler]: CheckboxState): CheckboxState {
+    return [state, () => { setIsPristine(false); toggler(); }];
+  }
+
+  const rawStates = map<CheckboxInitValues<K>, CheckboxStates<K>>(_useToggle, initialValues),
+      [isPristine, setIsPristine] = useState(true),
+      values = map<Pair, K>(head, filter<Pair, Pair[]>(([, [isSet]]) => isSet, toPairs(rawStates)));
```

Comment:
> Can you explain the process of how values is defined here? I understand map/head, but am unsure about the filter method here (what is [, [isSet]], and how is it applied to the rawStates). Thanks!

`toPairs` is a function which turns an object into a list of key-value pairs.  For example,
`toPairs({ a: 1, b: 2 })` results in `[['a', 1], ['b', 2]]`.  But in this case, all of the property values in `rawStates` are themselves lists (2-value tuples to be precise) because they're all return values from `_useToggle`.  So we have a triple-nested list being passed into `filter`.

For each item in that top-level list, `filter` applies the predicate that we've defined here as an arrow function.  That function uses destructing to pull out the only value in that structure that we're interested in, calls it `isSet`, and returns it.  When destructuring, you can ignore the values at list indexes you don't care about by simply writing the comma and no variable name to capture the value.  So for instance destructing with `[, x, , y]` would expect a list of at least four values, and would pull out the second of those values as the local variable `x` and the fourth of those values as the local variable `y`.  In our case, we are interested in the second value of that level of the structure (the filter predicate doesn't care about the name of the state).  

Now for that second value, which we do care about, we don't care about all of it.  This value is the third, deepest level of the triple-nested-list structure.  As I mentioned above, it's a `_useToggle` return value i.e. a `[state, toggleState]` structure.  The filter predicate function doesn't need the toggle function it only needs the state: whether or not this items _is set_.  Hence the name of the variable we destructure into.  So again, we are destructuring a list and only using part of it.  But in this case, the part we're ignoring is at the tail end rather than the front.  We could write this part as `[isSet, ]`, but that's not necessary.  We can just totally ignore that the rest of the list exists and write our destructuring as if that first value is all that's there.

So backing up, we use `toPairs` to turn the object into a list of key-value pairs.  We use filter to filter our the pairs whose state value is not currently `true` (using the arrow function to dig out the state value from each pair).  Then on the remaining pairs, we want their names - the keys.  So we map over them and call `head` on each one.  `head` is a function that retrieves the first value from a list. The first value of each pair is the key so that's what we want.

As I write this, it occurs to me that the filter predicate probably could have been written using ramda's `path` function instead (`path([1, 0])`).  That's certainly shorter; I'm not sure if it's clearer.  If you think it is I'd be happy to make that change.

Reply:
> As I write this, it occurs to me that the filter predicate probably could have been written using ramda's `path` function instead (`path([1, 0])`). That's certainly shorter; I'm not sure if it's clearer. If you think it is I'd be happy to make that change.

Thanks for such a detailed description, that was a big help. I didn't realize it was destructuring at work, so the code makes a lot more sense with your explanation. I would agree with you, that it may not be any clearer to use path in this case.